File: src/main/java/de/dheinrich/sandbox/ResourceLoading.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package de.dheinrich.sandbox;

import de.dheinrich.sandbox.meshstuff.MeshUtil;

import darwin.core.controls.*;
import darwin.core.gui.*;
import darwin.core.timing.*;
import darwin.geometrie.unpacked.*;
import darwin.renderer.geometrie.packed.RenderModel;
import darwin.renderer.geometrie.packed.RenderModel.RenderModelFactory;
import darwin.renderer.shader.*;
import darwin.resourcehandling.dependencies.annotation.*;
import darwin.resourcehandling.shader.ShaderLoader;
import darwin.resourcehandling.texture.*;
import darwin.util.math.composits.AABB;
import darwin.util.math.util.*;

import com.google.common.base.*;
import com.jogamp.opengl.util.texture.*;
import de.dheinrich.sandbox.meshstuff.*;
import javax.inject.Inject;
import javax.media.opengl.*;


/**
 *
 * @author Daniel Heinrich <dannynullzwo@gmail.com>
 */
public class ResourceLoading implements GLEventListener {

    @InjectBundle(files = {"sphere.frag", "sphere.vert"}, prefix = ShaderLoader.SHADER_PATH_PREFIX)
    private Shader shader;
    @InjectResource(file = "kinect.ctm")
    private Model[] kinect;
    @InjectResource(file = "out.png")
    private Texture rgb;
    
    @Inject
    private RenderModelFactory modeler;
    @Inject 
    private TextureUtil texUtil;
    
    private RenderModel model;
    private MatrixCache cache = new MatrixCache();
    private GameTime time = new GameTime();

    public static void main(String[] args) throws InstantiationException {
        boolean debug = true;

        Client client = Client.createClient(debug);
        ClientWindow window = new ClientWindow(1000, 500, false, client);
        window.startUp();

        final ResourceLoading a = client.addGLEventListener(ResourceLoading.class);

        FPSController vm = new FPSController();
        a.cache.setView(vm.getView());

        client.addMouseListener(new InputController(vm, null, a.cache));
        client.addKeyListener(vm);

        // a.time.addListener(vm.forCache(a.cache));
    }

    @Override
    public void init(GLAutoDrawable glad) {
//        glad.setGL(new DebugGL4(glad.getGL().getGL4()));

//        cache.addListener(shader);
        cache.addListener(shader);
        
        texUtil.setTexturePara(rgb, GL.GL_LINEAR, GL2.GL_CLAMP_TO_BORDER);
        float[] color = {0.1f, 0.1f, 0.1f, 0f};
        rgb.setTexParameterfv(glad.getGL(), GL2.GL_TEXTURE_BORDER_COLOR, color, 0);

        GL2GL3 gl = glad.getGL().getGL2GL3();
        gl.glClearColor(0.3f, 0.3f, 0.3f, 1);
        gl.glDisable(GL.GL_BLEND);
        gl.glEnable(GL.GL_DEPTH_TEST);
        
        gl.glFrontFace(GL.GL_CCW);
//        gl.glEnable(GL.GL_CULL_FACE);
        gl.glPolygonMode(GL.GL_FRONT, GL2.GL_LINE);

//        time.addListener(1, new StepListener() {
//            int frames;
//
//            @Override
//            public void update(int tickCount, float lerp, float tickTimeSpan) {
//                frames++;
//                if (tickCount > 0) {
//                    System.out.println(frames);
//                    frames = 0;
//                }
//            }
//        });
    }

    @Override
    public void display(GLAutoDrawable glad) {
        time.update();
        glad.getGL().glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT);

        if (shader.isInitialized()) {
            Optional<Sampler> diff = shader.getSampler("diffuse");
            diff.get().bindTexture(rgb);            
            
            if (model == null) {
//                NormalGenerator g = new NormalGenerator();
//                Mesh m = g.modifie(test[0].getMesh());
                model = modeler.create(kinect[0], shader);
                NormalGenerator ng = new NormalGenerator();
                Mesh m = ng.modifie(kinect[0].getMesh());
                model = modeler.create(new Model(m, null), shader);
                
                AABB a = MeshUtil.calcAABB(kinect[0].getMesh().getVertices());
                System.out.println(a);
//                for (Vector3 v : a.getCorners()) {
//                    GenericVector gv = new GenericVector(v.x, v.y, v.z, 1);
//                    Vector mult = cache.getViewProjection().mult(gv);
//                    mult.div(mult.getCoords()[3]);
//                    System.out.println(mult);
//                }
            }
            
            shader.updateUniformData();
            model.render();
        }
    }

    @Override
    public void reshape(GLAutoDrawable glad, int x, int y, int width, int height) {
        final float ratio = (float) width / Math.max(1, height);
        cache.getProjektion().perspective(60, ratio, 0.001f, 1000f); //ortho(-10, -10, -11110.1, 10, 10, 100);//
        cache.fireChange(MatType.PROJECTION);
    }

    @Override
    public void dispose(GLAutoDrawable glad) {
    }
}

File: src/main/scala/de/dheinrich/sandbox/scalarendering/ResourceLoading.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package de.dheinrich.sandbox.scalarendering

import com.google.common.base._
import com.jogamp.opengl.util.texture._
import darwin.core.controls._
import darwin.core.gui._
import darwin.core.timing._
import darwin.geometrie.unpacked._
import darwin.renderer.geometrie.packed.RenderModel
import darwin.renderer.shader._
import darwin.resourcehandling.dependencies.annotation._
import darwin.resourcehandling.shader.ShaderLoader
import darwin.resourcehandling.texture._
import darwin.util.math.base.vector.Vector3
import darwin.util.math.composits.AABB
import darwin.util.math.util._
import de.dheinrich.sandbox._
import de.dheinrich.sandbox.meshstuff._
import de.dheinrich.sandbox.observable.{TimeObservable, KeyEvent, KeyPressed, KeyObservable}
import javax.inject.Inject
import javax.media.opengl._
import rx.lang.scala.util._

/**
 *
 * @author Daniel Heinrich <dannynullzwo@gmail.com>
 */
object ResourceLoadingS {
  def main(args: Array[String]) {
    val debug: Boolean = true

    val client: Client = Client.createClient(debug)
    val window: ClientWindow = new ClientWindow(1000, 500, false, client)
    window.startUp

    val a = client.addGLEventListener(classOf[ResourceLoadingS])

    val fps: FPSController = new FPSController
    a.cache.setView(fps.getView)

    val time = TimeObservable.observeDelta(a.time)

    //val events = KeyObservable.observe(client) merge MouseObservable.observe(client)
    val keys = KeyObservable.observe(client)

    val until = keys.buffer(() => time.map(t => Closing())).map(_.groupBy(_.code.toChar).toMap.mapValues(_.last))


    def isPressed(c: Char) = (_: Map[Char, KeyEvent]).get(c).filter(_.isInstanceOf[KeyPressed]).isDefined

    //    val forward = until.map(isPressed('w'))
    //    val back = until.map(isPressed('s'))
    //    val left = until.map(isPressed('a'))
    //    val right = until.map(isPressed('d'))

    def conv(b: Boolean): Int = {
      return if (b) -1 else 1
    }
    val speed = 10
    val dirSwitch = (t: (Boolean, Boolean)) => if (t._1 || t._2) conv(t._1) * speed else 0

    //    val z = (forward zip back).map(dirSwitch)
    //    val x = (left zip right).map(dirSwitch)
    //
    //    val direction = (x zip z) map (t => new Vector3(t._1, 0, t._2))


    val direction = for (m <- until) yield {
      val f = isPressed('w')(m)
      val b = isPressed('s')(m)
      val l = isPressed('a')(m)
      val r = isPressed('d')(m)

      val z = dirSwitch((f, b))
      val x = dirSwitch((l, r))
      new Vector3(x, 0, z)
    }

    direction.subscribe({
      trans =>
        fps.position.add(fps.rotation.getRotationMatrix.fastMult(trans))
        fps.resetInverse
        a.cache.fireChange(MatType.VIEW)
    }, _.printStackTrace())

    client.addMouseListener(new InputController(fps, null, a.cache))
    client.addKeyListener(fps)
  }
}

class ResourceLoadingS extends GLEventListener {
  @InjectBundle(files = Array("sphere.frag", "sphere.vert"), prefix = ShaderLoader.SHADER_PATH_PREFIX)
  private var shader: Shader = null

  @InjectResource(file = "kinect.ctm")
  private var kinect: Array[Model] = null

  @InjectResource(file = "out.png")
  private var rgb: Texture = null

  @Inject
  private var modeler: RenderModel.RenderModelFactory = null
  @Inject
  private var texUtil: TextureUtil = null

  private var model: RenderModel = null
  var cache: MatrixCache = new MatrixCache
  var time: GameTime = new GameTime

  def init(glad: GLAutoDrawable) {
    import GL._
    import GL2GL3._

    cache.addListener(shader)
    texUtil.setTexturePara(rgb, GL_LINEAR, GL_CLAMP_TO_BORDER)

    val color: Array[Float] = Array(0.1f, 0.1f, 0.1f, 0f)
    rgb.setTexParameterfv(glad.getGL, GL_TEXTURE_BORDER_COLOR, color, 0)

    val gl = glad.getGL().getGL2GL3();
    {
      import gl._
      glClearColor(0.3f, 0.3f, 0.3f, 1)
      glDisable(GL_BLEND)
      glEnable(GL_DEPTH_TEST)
      glFrontFace(GL_CCW)
      glPolygonMode(GL_FRONT, GL_LINE)
    }
  }

  def display(glad: GLAutoDrawable) {
    time.update
    glad.getGL.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT)

    if (shader.isInitialized) {
      val diff: Optional[Sampler] = shader.getSampler("diffuse")
      diff.get.bindTexture(rgb)
      if (model == null) {
        model = modeler.create(kinect(0), shader)
        val ng: NormalGenerator = new NormalGenerator
        val m: Mesh = ng.modifie(kinect(0).getMesh)
        model = modeler.create(new Model(m, null), shader)
        val a: AABB = MeshUtil.calcAABB(kinect(0).getMesh.getVertices)
        System.out.println(a)
      }
      shader.updateUniformData
      model.render
    }
  }

  def reshape(glad: GLAutoDrawable, x: Int, y: Int, width: Int, height: Int) {
    val ratio: Float = width.asInstanceOf[Float] / Math.max(1, height)
    cache.getProjektion.perspective(60, ratio, 0.001f, 1000f)
    cache.fireChange(MatType.PROJECTION)
  }

  def dispose(glad: GLAutoDrawable) {
  }
}